core: Add list() interpolation function

jen20 opened the pull request:
The list() interpolation function provides a way to add support for list literals to HIL without having to invent new syntax for it and modify the HIL parser.

It presents as a function, thus:

    - list() -> []
    - list("a") -> ["a"]
    - list("a", "b") -> ["a", "b"]

Thanks to @wr0ngway for the idea of this approach, fixes #7460.

Assuming this is the route we wish to follow, we should make it work with types other than strings - the initial quick implementation is to get a feel for using a function rather than syntax.

cc @phinze, @mitchellh, @jbardin

phinze commented:
Direction looks good. +1 for this as a stop-gap solution until we can parse non-primitives directly in HIL.

jen20 commented:
I've cleaned this up:

- added a test to verify behaviour if a non-string is passed
- removed the type checking code (which would never be hit) in favour of letting the type checker do its work
- fixed formatting issues

I'll remove the WIP now - cc @phinze.

phinze commented:
LGTM for strings. Agreed we should make it work for lists and maps as well.

jbardin commented:
Expanded the list interpolation to accept lists and maps, and the concat function to accept lists of lists and lists of maps. 

This now also fixes #7146

phinze commented on config/interpolate_funcs.go:
```diff
@@ -218,44 +258,52 @@ func interpolationFuncCoalesce() ast.Function {
+				switch arg := arg.(type) {
+				case string:
+					// we also allow bare strings for backwards compatibility
+					outputList = append(outputList, ast.Variable{Type: ast.TypeString, Value: arg})
```
Let's nuke the BC here. I believe we've already gone talking about this case like it's gone! 😀

phinze commented on config/interpolate_funcs_test.go:
```diff
@@ -5,13 +5,114 @@ import (
+			},
+		},
+	})
+}
```
👏

phinze commented on config/interpolate_funcs.go:
```diff
@@ -218,44 +258,52 @@ func interpolationFuncCoalesce() ast.Function {
+				firstType := outputList[0].Type
+				for _, v := range outputList[1:] {
+					if v.Type != firstType {
+						return nil, fmt.Errorf("unexpected type %s in list of type %s", v.Type, firstType)
```
I believe "printable" is a thing for types to make them more error string friendly.

jbardin replied:
I wasn't sure, because we also stringer-ify them.  The Printable() output is all lower case, but there's no grammar-correct way to remove the stutter from that output (not that this is much better :shrug:)

phinze replied:
Ah gotchya - yeah I was really just making sure we are treating this as a user-facing error message and making sure it's reasonably decipherable from out front. I'll defer to whatever you think works best here. 👍

phinze commented:
Nice work @jbardin!! A few inline nits, but this looks super solid overall.

jbardin commented:
Used Printable, since it will usually make a better user-facing error.
Squashed the minor changes.

phinze commented:
LGTM
